src/Wasp.php

```diff
@@ -0,0 +1,239 @@
+use Symfony\Component\EventDispatcher\EventDispatcher;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Console\Logger\ConsoleLogger;
+use Symfony\Component\Console\Input\InputOption;
+use Symfony\Component\Console\Input\InputDefinition;
+use OomphInc\WASP\Input\PartialInputDefinition;
+use Symfony\Component\Console\ConsoleEvents;
+use OomphInc\WASP\Event\Events;
+use Closure;
+
+class Wasp {
+
+	use ServiceContainerTrait;
+
+	const NAME = 'wasp';
+	const VERSION = 'beta';
+	const COMPOSER_TYPE = 'wasp-plugin';
+
+	protected $application;
+	protected $plugins = [];
+	protected $globalInputDefinition;
+
+	/**
+	 * @param array  $services default services
+	 * @param string $lockFile contents of composer lock file for the project
+	 */
+	public function __construct(array $services = [], $lockFile = null) {
+		$this->handleServiceDefaults($services);
+
+		// create the application
+		$application = new Application(static::NAME, static::VERSION);
+		// add global options
+		$application->getDefinition()->addOptions($this->getGlobalOptions());
+		// add commands
+		$application->add(new Command\Generate($this));
+		// do not auto exit upon command completion
+		$application->setAutoExit(false);
+		$this->setApplication($application);
+
+		// bind input just to global params for early processing
+		$input = $this->getService('input');
+		$input->bind($this->getGlobalInputDefinition());
+		// create a bound closure to call the protected method that sets up output verbosity and format
+		$configureIO = Closure::bind(function($input, $output) {
+			$this->configureIO($input, $output);
+		}, $application, $application);
+		$configureIO($input, $this->getService('output'));
+
+		// optional files to include
+		foreach ($input->getOption('include') as $file) {
+			if (!is_readable($file)) {
+				throw new RuntimeException("File does not exist or is not readable: $file");
+			}
+			require_once $file;
+		}
+
+		// parse lock file and initialize plugins
+		if ($lockFile) {
+			if ($input->getOption('no-plugins')) {
+				$this->getService('logger')->debug('Skipping plugins due to --no-plugins flag');
+			} else {
+				$this->initializePluginsFromLock($lockFile, $input->getOption('disable-plugin'));
+			}
+		}
+
+		$this->getService('dispatcher')->dispatch(Events::PLUGINS_LOADED);
+	}
+
+	/**
+	 * Get the default types of services that may be accepted.
+	 * @return array [service => class type, ...]
+	 */
+	protected function getDefaultServiceTypes() {
+		return [
+			'input' => 'Symfony\Component\Console\Input\InputInterface',
+			'output' => 'Symfony\Component\Console\Output\OutputInterface',
+			'logger' => 'Psr\Log\LoggerInterface',
+			'dispatcher' => 'Symfony\Component\EventDispatcher\EventDispatcherInterface',
+			'filesystem' => __NAMESPACE__ . '\\FileSystem\\FileSystemInterface',
+			'stdin' => __NAMESPACE__ . '\\Input\\StdInInterface',
+			'linter' => __NAMESPACE__ . '\\Linter\\LinterInterface',
```

Comment:
Do you really need to escape these backslashes?

Reply:
technically no, and clearly it's inconsistent with the non concatenated ones above D: